### PR TITLE
feat: swap to/from RUNE

### PR DESCRIPTION
**Pull request by asamere**
Adds support for swapping to/from RUNE, the native asset of Thorchain. Also supporting Thorchain synth assets would be a small stretch.

### CODE REVIEW

File: packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts [hunk 1]
```diff
@@ -1,16 +1,20 @@
...
 import { THOR_TRADE_FEE_MULTIPLIERS } from '../constants'
 import { getPriceRatio } from '../getPriceRatio/getPriceRatio'
 import { thorService } from '../thorService'
 
+// TODO rename this estimateOutboundFee or similar
 export const estimateTradeFee = async (
   deps: ThorchainSwapperDeps,
   buyAsset: Asset,
 ): Promise<string> => {
+  if (buyAsset.assetId == thorchainAssetId) {
```
Comment (asamere) [inline packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts:15]: add `isRune(AssetId)` utility

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts [hunk 2]
```diff
@@ -174,6 +174,91 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.ThorchainMa
...
+      // TODO memo validation
+      if (!memo) throw new Error('ThorchainChainAdapter: value is required')
+      if (!value) throw new Error('ThorchainChainAdapter: value is required')
+
+      const addressNList = toAddressNList(bip44Params)
+      const from = await this.getAddress({ bip44Params, wallet })
+
+      const account = await this.getAccount(from)
+
+      // 0.02 RUNE is automatically charged on outbound transactions
```
Comment (asamere) [inline packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:200]: this is kinda crazy

File: packages/swapper/src/adapters.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
+import { ChainId } from '@shapeshiftoss/caip'
```
Comment (asamere) [inline packages/swapper/src/adapters.ts:1]: gonna move this to chain-adapters

File: packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts (hunk 1 above)
Comment (asamere) [reply packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts:15]: use isRune

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo/makeSwapMemo.ts [hunk 4]
```diff
@@ -19,7 +21,8 @@ export const makeSwapMemo = ({
   destinationAddress: string
   limit: string
 }): string => {
-  const thorId = adapters.assetIdToPoolAssetId({ assetId: buyAssetId })
+  const isRune = buyAssetId === thorchainAssetId
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo/makeSwapMemo.ts:24]: can we use our little function above here?

https://github.com/shapeshift/lib/pull/1030/files#diff-beaf016013438f20332d1ae30ebda43adf09fa660169bfb5c1edde2d3ff36f4fR3

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts [hunk 5]
```diff
@@ -174,6 +175,90 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.ThorchainMa
     }
   }
 
+  /* MsgDeposit is used for thorchain swap/lp operations */
+  async buildDepositTransaction(tx: BuildDepositTxInput): Promise<{ txToSign: ThorchainSignTx }> {
+    try {
+      const {
+        wallet,
+        bip44Params = this.defaultBIP44Params,
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:183]: Probably shouldn't need to use default bip44Params anymore - these are actually coming from the base adapter but we should be able to send them as part of the input, see https://github.com/shapeshift/lib/pull/1021

File: packages/chain-adapters/src/cosmossdk/thorchain/types.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { HDWallet } from '@shapeshiftoss/hdwallet-core'
+import { BIP44Params } from '@shapeshiftoss/types'
+
+export type BuildDepositTxInput = {
+  memo: string
+  value: string
+  wallet: HDWallet
+  gas: string
+  fee: string
+  bip44Params?: BIP44Params // TODO mandatory for multi account
```
Comment (gomesalexandre) [inline packages/chain-adapters/src/cosmossdk/thorchain/types.ts:10]: Non-optional

**Comment by gomesalexandre** [comment]
Note: this requires a few rough edges of required BIP44Params to be tackled in https://github.com/shapeshift/lib/pull/1021

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts [hunk 7]
```diff
@@ -14,6 +14,7 @@ import {
 import { toAddressNList } from '../../utils'
 import { bnOrZero } from '../../utils/bignumber'
 import { ChainAdapterArgs, CosmosSdkBaseAdapter } from '../CosmosSdkBaseAdapter'
+import { BuildDepositTxInput } from './types'
 
 // static automatic outbound fee as defined by: https://daemon.thorchain.shapeshift.com/thorchain/constants
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:19]: dead link

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts [hunk 8]
```diff
@@ -97,13 +102,17 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.ThorchainMa
...
-        bip44Params = this.defaultBIP44Params,
+        bip44Params,
         chainSpecific: { gas, fee },
         sendMax = false,
         value,
         memo = '',
       } = tx
 
+      if (!bip44Params) {
+        throw new Error('bip44Params required in buildSendTransaction input')
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:113]:
```suggestion
        throw new Error('bip44Params required in signTransaction input')
```

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts [hunk 9]
```diff
@@ -174,6 +183,90 @@ export class ChainAdapter extends CosmosSdkBaseAdapter<KnownChainIds.ThorchainMa
...
+        bip44Params = this.defaultBIP44Params,
+        gas,
+        fee,
+        sendMax = false,
+        value,
+        memo,
+      } = tx
+
+      // TODO memo validation
+      if (!memo) throw new Error('ThorchainChainAdapter: value is required')
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:200]:
```suggestion
      if (!memo) throw new Error('ThorchainChainAdapter: memo is required')
```

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1030** <a href="https://app.graphite.dev/github/pr/shapeshift/lib/1030" target="_blank">Graphite</a>  👈
    * **PR #1037** <a href="https://app.graphite.dev/github/pr/shapeshift/lib/1037" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/lib/1030?utm_source=stack-comment).

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts (hunk 5 above)
Comment (gomesalexandre) [reply packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:191]: Disregard the previous comment, let's get this in as is, defaultBip44Params will be removed everywhere in #1021

File: packages/swapper/src/swappers/thorchain/utils/makeSwapMemo/makeSwapMemo.ts [hunk 10]
```diff
@@ -1,11 +1,13 @@
...
+import { adapters, thorchainAssetId } from '@shapeshiftoss/caip'
 
 import { SwapError, SwapErrorTypes } from '../../../../api'
 import { THORCHAIN_AFFILIATE_BIPS, THORCHAIN_AFFILIATE_NAME } from '../constants'
 
 // BTC (and likely other utxo coins) can only support up to 80 character memos
 const MAX_LENGTH = 80
 
+// use symbol of thor Asset?
+const runeThorId = 'RUNE'
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/thorchain/utils/makeSwapMemo/makeSwapMemo.ts:10]: @asamere Do we want to add this guy to `thorPoolIdAssetIdSymbolMap` ?

File: packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts [hunk 11]
```diff
@@ -5,12 +5,17 @@ import { SwapError, SwapErrorTypes } from '../../../../api'
...
+import { isRune } from '../isRune/isRune'
 import { thorService } from '../thorService'
 
+// TODO rename this estimateOutboundFee or similar
 export const estimateTradeFee = async (
   deps: ThorchainSwapperDeps,
   buyAsset: Asset,
 ): Promise<string> => {
+  if (isRune(buyAsset.assetId)) {
+    return '0.02' // todo - read from TC mimir/constants
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/thorchain/utils/estimateTradeFee/estimateTradeFee.ts:17]: @asamere Do we want to expose `NativeTransactionFee` (https://thornode.ninerealms.com/thorchain/constants) in unchained, or is the plan to query a thor node endpoint directly and pass this one over?

File: packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts (hunk 7 above)
Comment (gomesalexandre) [reply packages/chain-adapters/src/cosmossdk/thorchain/ThorchainChainAdapter.ts:19]: https://github.com/shapeshift/lib/pull/1030/files#r979792342 - unless (until?) this is exposed by `https://daemon.thorchain.shapeshift.com`, going to move this comment over to the `thornode.ninerealms.com` endpoint

File: packages/chain-adapters/src/cosmossdk/thorchain/types.ts [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+export type BuildDepositTxInput = {
+  memo: string
+  value: string
+  wallet: HDWallet
+  gas: string
+  fee: string
+  bip44Params: BIP44Params
+  // for synth support:
+  // asset?: AssetId // synths/ibc?
```
Comment (0xdef1cafe) [inline packages/chain-adapters/src/cosmossdk/thorchain/types.ts:12]: remove if we're not using yet or plan to?

File: packages/swapper/src/adapters.ts [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  ChainAdapterManager,
+  cosmos,
+  ethereum,
+  thorchain,
+} from '@shapeshiftoss/chain-adapters'
+import { KnownChainIds } from '@shapeshiftoss/types'
+import * as unchained from '@shapeshiftoss/unchained-client'
+import dotenv from 'dotenv'
+
+dotenv.config()
```
Comment (0xdef1cafe) [inline packages/swapper/src/adapters.ts:14]: this concerns me, we'd have to (and should) exclude this file from the build

File: packages/swapper/src/api.ts [hunk 14]
```diff
@@ -64,6 +64,7 @@ export type SupportedSellAssetsInput = {
 }
 
 type CommonTradeInput = {
+  bip44Params?: BIP44Params
```
Comment (0xdef1cafe) [inline packages/swapper/src/api.ts:67]: should this be required?

Comment (gomesalexandre) [reply packages/swapper/src/api.ts:67]: It becomes required on the stacked PR (https://github.com/shapeshift/lib/pull/1037) - we can't make this required until we remove `sellAssetAccountNumber` or a bunch of places will be sad

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts [hunk 15]
```diff
@@ -49,39 +49,80 @@ export const getTradeRate = async (
...
-  const sellAssetPoolData = {
-    assetBalance: bn(sellPool.assetDepth),
-    runeBalance: bn(sellPool.runeDepth),
-  }
-  const buyAssetPoolData = {
-    assetBalance: bn(buyPool.assetDepth),
-    runeBalance: bn(buyPool.runeDepth),
+  const doubleSwap = buyPool != undefined && sellPool != undefined
+  const toRune = buyAssetId === thorchainAssetId
+  let outputAmountBase8: BN
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts:96]: iife?

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts [hunk 16]
```diff
@@ -49,39 +49,80 @@ export const getTradeRate = async (
...
+          runeBalance: bn(sellPool.runeDepth),
+        }
+      : buyPool && {
+          assetBalance: bn(buyPool.assetDepth),
+          runeBalance: bn(buyPool.runeDepth),
+        }
+    if (!poolData) {
+      throw new SwapError(`[getPriceRatio]: No pool data for rune swap`, {
+        code: SwapErrorTypes.POOL_NOT_FOUND,
+        fn: 'getPriceRatio',
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts:120]:
```suggestion
        fn: 'getTradeRate',
```

File: packages/swapper/src/swappers/thorchain/utils/getUsdRate/getUsdRate.ts [hunk 17]
```diff
@@ -25,10 +35,22 @@ export const getUsdRate = async ({
...
+    const rate: string | undefined = (() => {
+      if (isRune(assetId)) {
+        const bnRate = bnOrZero(responseData?.assetPrice)
+        if (bnRate.isZero()) {
+          throw new SwapError('[getUsdRate]: cannot invert rate zero', {
+            code: SwapErrorTypes.USD_RATE_FAILED,
+          })
+        }
+        const inverseRate = bn(1).div(bnRate)
+        return inverseRate.toFixed(8)
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/getUsdRate/getUsdRate.ts:47]:
```suggestion
        return inverseRate.toFixed(THOR_PRECISION)
```

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts [hunk 18]
```diff
@@ -49,39 +49,80 @@ export const getTradeRate = async (
   const buyPoolId = adapters.assetIdToPoolAssetId({ assetId: buyAssetId })
   const sellPoolId = adapters.assetIdToPoolAssetId({ assetId: sellAsset.assetId })
 
-  if (!buyPoolId || !sellPoolId)
-    throw new SwapError(`[getPriceRatio]: No thorchain pool found`, {
+  if (!buyPoolId && buyAssetId !== thorchainAssetId) {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts:52]:
```suggestion
  if (!buyPoolId && !isRune(buyAssetId)) {
```

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts [hunk 19]
```diff
@@ -49,39 +49,80 @@ export const getTradeRate = async (
...
+    throw new SwapError(`[getTradeRate]: No buyPoolId for asset ${buyAssetId}`, {
       code: SwapErrorTypes.POOL_NOT_FOUND,
       fn: 'getPriceRatio',
-      details: { buyPoolId, sellPoolId },
+      details: { buyAssetId },
     })
+  }
+
+  if (!sellPoolId && sellAsset.assetId !== thorchainAssetId) {
```
Comment (0xdef1cafe) [inline packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts:60]:
```suggestion
  if (!sellPoolId && !isRune(sellAsset.assetId)) {
```

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts (hunk 18 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts (hunk 19 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

File: packages/chain-adapters/src/cosmossdk/thorchain/types.ts (hunk 12 above)
Comment (gomesalexandre) [reply packages/chain-adapters/src/cosmossdk/thorchain/types.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts (hunk 15 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

File: packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts (hunk 16 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/utils/getTradeRate/getTradeRate.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

File: packages/swapper/src/swappers/thorchain/utils/getUsdRate/getUsdRate.ts (hunk 17 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/thorchain/utils/getUsdRate/getUsdRate.ts]: 07c41f9764d6207cd5323bbef93f8da0eb1d912c

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**